### ISSUE TITLE
Chaining Operations - Misunderstanding?

### ISSUE THREAD

**Issue by ShreyasSkandan** [pytorch/pytorch#13438]
Hi,

I'm having a little trouble understanding why I'm getting different results for the following operations that seem identical to me. I've been smacking my brain over this for a while now so any thoughts will be extremely helpful.

Let's say I have the following Tensors
Tensor_A: dimension => [batch_size, 1, 250, 500]
Tensor_B: dimension => [batch_size, 1, 250, 500]

I'm trying to find the Mean Absolute Error of the two tensors, but there's an additional caveat to do this over only a certain set of pixels for each image. So i have a Mask that I got from checking to see if some matrix had elements greater than 0,

Tensor_Mask = (SomeTensor > 0).detach()
Tensor_Mask: dimension => [batch_size, 1, 250, 500]

My calculations are as follows:

Difference_AB = Tensor_A - Tensor_B
MaskedError = Difference_AB[Tensor_Mask]
MAE_Scalar = MaskedError.abs().mean()

Let's take a more elaborate way to compute the same MAE_Scalar

Difference_AB = Tensor_A - Tensor_B
MaskedDiff = torch.mul(Difference_AB, Tensor_Mask.float())
MaskedErrAbs = torch.abs(MaskedDiff)
MaskedAverageError = torch.mean(MaskedErrAbs)

The two values MAE_Scalar and MaskedAverageError don't seem to be the same. 

Any thoughts?

Am I doing something extremely silly..?

**Comment by SsnL** [comment]
after masking your tensor have less elements in method (1) and more elements with a bunch of zeros in method (2). the denominator is different.
